// @mxw/next/src/helper/axeshelper.ts
import { wom, ManualVisual, Section } from 'maxwhere';

export default class AxesHelper {
  manualObj: ManualVisual;
  constructor(size: number) {
    let section: Section = {
      material_name: 'Physics/DebugDraw/VertexDriven',
      renderop: 'linelist',
      geometry: [
        // x axis

        {
          type: 'position',
          arg: { x: 0, y: 0, z: 0 },
        },
        {
          type: 'color',
          arg: { r: 1, g: 0, b: 0, a: 1 },
        },

        {
          type: 'position',
          arg: { x: size, y: 0, z: 0 },
        },
        {
          type: 'color',
          arg: { r: 1, g: 0, b: 0, a: 1 },
        },

        // y axis

        {
          type: 'position',
          arg: { x: 0, y: 0, z: 0 },
        },
        {
          type: 'color',
          arg: { r: 0, g: 1, b: 0, a: 1 },
        },

        {
          type: 'position',
          arg: { x: 0, y: size, z: 0 },
        },
        {
          type: 'color',
          arg: { r: 0, g: 1, b: 0, a: 1 },
        },

        // z axis

        {
          type: 'position',
          arg: { x: 0, y: 0, z: 0 },
        },
        {
          type: 'color',
          arg: { r: 0, g: 0, b: 1, a: 1 },
        },

        {
          type: 'position',
          arg: { x: 0, y: 0, z: size },
        },
        {
          type: 'color',
          arg: { r: 0, g: 0, b: 1, a: 1 },
        },
      ],
    };

    this.manualObj = wom.create('manualvisual', {
      sections: [section],
      // to make sure its above a gridhelper if there is any
      position: { x: 0, y: 0.001, z: 0 },
    });

    wom.render(this.manualObj);
  }

  clear() {
    this.manualObj.clear();
  }
}
