// @mxw/three-utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// Collection of 3D Utility functions
const THREE = __importStar(require("@mxw/three"));
/**
 * Converts Three.js Quatrenion to MaxWhere quaterion.
 * @param q
 * @returns
 */
function QuatToMw(q) {
    return {
        x: q.x,
        y: q.y,
        z: q.z,
        w: q.w,
    };
}
/**
 * Converts Three.js Vector3 to MaxWhere vector.
 * @param v
 * @returns
 */
function V3ToMw(v) {
    return {
        x: v.x,
        y: v.y,
        z: v.z,
    };
}
/**
 * Converts MaxWhere vector to Three.js Vector3
 * @param pos
 * @returns
 */
function V3ToThree(pos) {
    return new THREE.Vector3(pos.x, pos.y, pos.z);
}
/**
 * Convert MaxWhere quaternion to Three.js Quaternion
 * @param ori
 * @returns
 */
function QuatToThree(ori) {
    return new THREE.Quaternion(ori.x, ori.y, ori.z, ori.w);
}
/**
 * Rotates a MaxWhere quaternion on an axis by a defined angle
 * @param quaterion initial quaternion
 * @param axis A normalized vector representing the axis
 * @param angle angle in degree
 * @returns
 */
function rotateQuaternionOnAxis(quaterion, axis, angle) {
    let baseOri = QuatToThree(quaterion);
    let rotateOffset = new THREE.Quaternion().setFromAxisAngle(V3ToThree(axis), THREE.MathUtils.degToRad(angle));
    let newOri = QuatToMw(new THREE.Quaternion().multiplyQuaternions(baseOri, rotateOffset));
    return newOri;
}
/**
 * Look at a direction with an object.
 * @param forward forward direction at identity quaternion
 * @param up up direction at identity quaternion
 * @param dir direction vector to rotate to
 * @returns
 */
function lookAtDir(forward, up, dir) {
    let rotationMatrix = new THREE.Matrix4().lookAt(dir, new THREE.Vector3(0, 0, 0), up);
    let ori = QuatToMw(new THREE.Quaternion()
        .setFromRotationMatrix(rotationMatrix)
        .multiply(new THREE.Quaternion().setFromUnitVectors(new THREE.Vector3(0, 0, -1), forward)));
    return ori;
}
/**
 * Look at a direction with an object.
 * @param forward forward direction at identity quaternion
 * @param up up direction at identity quaternion
 * @param dir direction vector to rotate to
 * @returns
 */
function lookAt(fromPos, toPos, forward, up) {
    let rotationMatrix = new THREE.Matrix4().lookAt(fromPos, toPos, up || new THREE.Vector3(0, 1, 0));
    let ori = new THREE.Quaternion()
        .setFromRotationMatrix(rotationMatrix)
        .multiply(new THREE.Quaternion().setFromUnitVectors(new THREE.Vector3(0, 0, -1), forward || new THREE.Vector3(0, 0, 1)));
    return ori;
}
function worldToLocal(pos, matrix) {
    return pos.clone().applyMatrix4(matrix.clone().invert());
}
/**
 * Translates MaxWhere vector on a direction by a defined distance
 * @param position Initial position
 * @param direction direction vector
 * @param distance
 * @returns
 */
function translatePositionOnDirection(position, direction, distance) {
    return V3ToMw(V3ToThree(direction).multiplyScalar(distance).add(V3ToThree(position)));
}
function V3Flip(pos) {
    return new THREE.Vector3(1 / pos.x, 1 / pos.y, 1 / pos.z);
}
/**
 * Converts quaternion to angle and axis
 * @param quat quaternion
 * @returns AngleAxis object
 */
function getAngleAxisFromQuaternion(quat) {
    let axis = { x: 0, y: 0, z: 0 };
    if (quat.w > 1)
        quat.normalize(); // if w>1 acos and sqrt will produce errors, this cant happen if quaternion is normalised
    let angle = THREE.MathUtils.radToDeg(2 * Math.acos(quat.w));
    let s = Math.sqrt(1 - quat.w * quat.w); // assuming quaternion normalised then w is less than 1, so term always positive.
    if (s < 0.001) {
        // test to avoid divide by zero, s is always positive due to sqrt
        // if s close to zero then direction of axis not important
        axis.x = quat.x; // if it is important that axis is normalised then replace with x=1; y=z=0;
        axis.y = quat.y;
        axis.z = quat.z;
    }
    else {
        axis.x = quat.x / s; // normalise axis
        axis.y = quat.y / s;
        axis.z = quat.z / s;
    }
    axis.x = Math.round(axis.x * 100) / 100;
    axis.y = Math.round(axis.y * 100) / 100;
    axis.z = Math.round(axis.z * 100) / 100;
    return { axis: axis, angle: angle };
}
exports.default = {
    QuatToMw,
    V3ToMw,
    V3ToThree,
    QuatToThree,
    translatePositionOnDirection,
    V3Flip,
    rotateQuaternionOnAxis,
    lookAt: lookAt,
    lookAtDir: lookAtDir,
    getAngleAxisFromQuaternion,
    worldToLocal: worldToLocal,
};
