// @mxw/next/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.powerUp = exports.womAsync = exports.camera = exports.Gizmo = exports.ArrowHelper = exports.PolarGridHelper = exports.GridHelper = exports.AxesHelper = exports.CSG = exports.GeometryBuilder = exports.MeshTransform = exports.Transform = void 0;
var Transform_1 = require("./wrappers/Transform");
Object.defineProperty(exports, "Transform", { enumerable: true, get: function () { return __importDefault(Transform_1).default; } });
var MeshTransform_1 = require("./wrappers/MeshTransform");
Object.defineProperty(exports, "MeshTransform", { enumerable: true, get: function () { return __importDefault(MeshTransform_1).default; } });
var geometryBuilder_1 = require("./wrappers/geometryBuilder");
Object.defineProperty(exports, "GeometryBuilder", { enumerable: true, get: function () { return __importDefault(geometryBuilder_1).default; } });
var CSG_1 = require("./CSG/CSG");
Object.defineProperty(exports, "CSG", { enumerable: true, get: function () { return __importDefault(CSG_1).default; } });
var axeshelper_1 = require("./helper/axeshelper");
Object.defineProperty(exports, "AxesHelper", { enumerable: true, get: function () { return __importDefault(axeshelper_1).default; } });
var gridHelper_1 = require("./helper/gridHelper");
Object.defineProperty(exports, "GridHelper", { enumerable: true, get: function () { return __importDefault(gridHelper_1).default; } });
var polarGridHelper_1 = require("./helper/polarGridHelper");
Object.defineProperty(exports, "PolarGridHelper", { enumerable: true, get: function () { return __importDefault(polarGridHelper_1).default; } });
var arrowHelper_1 = require("./helper/arrowHelper");
Object.defineProperty(exports, "ArrowHelper", { enumerable: true, get: function () { return __importDefault(arrowHelper_1).default; } });
var index_1 = require("./gizmo/index");
Object.defineProperty(exports, "Gizmo", { enumerable: true, get: function () { return __importDefault(index_1).default; } });
var camera_1 = require("./wrappers/camera");
Object.defineProperty(exports, "camera", { enumerable: true, get: function () { return __importDefault(camera_1).default; } });
var womAsync_module_1 = require("./wrappers/womAsync.module");
Object.defineProperty(exports, "womAsync", { enumerable: true, get: function () { return __importDefault(womAsync_module_1).default; } });
var index_2 = require("./powerUp/index");
Object.defineProperty(exports, "powerUp", { enumerable: true, get: function () { return __importDefault(index_2).default; } });
exports.utils = __importStar(require("./utils/index"));
