// @mxw/next/dist/helper/arrowHelper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const geometryBuilder_1 = __importDefault(require("../wrappers/geometryBuilder"));
const THREE = __importStar(require("@mxw/three"));
const three_utils_1 = __importDefault(require("@mxw/three-utils"));
const maxwhere_1 = require("maxwhere");
class ArrowHelper {
    constructor(dir, origin, length = 100, material_name = 'Physics/DebugDraw/VertexDriven', color = { r: 1, g: 0, b: 0, a: 1 }, headLength = 0.2 * length, headWidth = 0.2 * headLength) {
        this.dir = dir;
        this.color = color;
        this.origin = origin;
        this.length = length;
        this.material_name = material_name;
        this.headLength = headLength;
        this.headWidth = headWidth;
        this.calculateSections();
        this.arrowObj = maxwhere_1.wom.create('manualvisual', {
            sections: [this.lineSection, this.coneSection],
        });
        maxwhere_1.wom.render(this.arrowObj);
    }
    updateVisual() {
        this.calculateSections();
        this.arrowObj.updateSection(0, this.lineSection.geometry),
            this.arrowObj.updateSection(1, this.coneSection.geometry);
    }
    calculateSections() {
        let endPoint = this.origin
            .clone()
            .add(this.dir.clone().multiplyScalar(this.length));
        this.lineSection = {
            material_name: this.material_name,
            renderop: 'linelist',
            geometry: [
                { type: 'color', arg: this.color },
                { type: 'position', arg: three_utils_1.default.V3ToMw(this.origin) },
                {
                    type: 'position',
                    arg: three_utils_1.default.V3ToMw(endPoint),
                },
            ],
        };
        let coneGeometry = new THREE.ConeGeometry(this.headWidth, this.headLength);
        coneGeometry.applyQuaternion(new THREE.Quaternion().setFromUnitVectors(new THREE.Vector3(0, 1, 0), this.dir));
        coneGeometry.translate(endPoint.x, endPoint.y, endPoint.z);
        this.coneSection = geometryBuilder_1.default.getFacesFromGeometry(coneGeometry, this.material_name, this.color);
    }
    clear() {
        this.arrowObj.clear();
    }
}
exports.default = ArrowHelper;
