// @mxw/next/dist/helper/axeshelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const maxwhere_1 = require("maxwhere");
class AxesHelper {
    constructor(size) {
        let section = {
            material_name: 'Physics/DebugDraw/VertexDriven',
            renderop: 'linelist',
            geometry: [
                // x axis
                {
                    type: 'position',
                    arg: { x: 0, y: 0, z: 0 },
                },
                {
                    type: 'color',
                    arg: { r: 1, g: 0, b: 0, a: 1 },
                },
                {
                    type: 'position',
                    arg: { x: size, y: 0, z: 0 },
                },
                {
                    type: 'color',
                    arg: { r: 1, g: 0, b: 0, a: 1 },
                },
                // y axis
                {
                    type: 'position',
                    arg: { x: 0, y: 0, z: 0 },
                },
                {
                    type: 'color',
                    arg: { r: 0, g: 1, b: 0, a: 1 },
                },
                {
                    type: 'position',
                    arg: { x: 0, y: size, z: 0 },
                },
                {
                    type: 'color',
                    arg: { r: 0, g: 1, b: 0, a: 1 },
                },
                // z axis
                {
                    type: 'position',
                    arg: { x: 0, y: 0, z: 0 },
                },
                {
                    type: 'color',
                    arg: { r: 0, g: 0, b: 1, a: 1 },
                },
                {
                    type: 'position',
                    arg: { x: 0, y: 0, z: size },
                },
                {
                    type: 'color',
                    arg: { r: 0, g: 0, b: 1, a: 1 },
                },
            ],
        };
        this.manualObj = maxwhere_1.wom.create('manualvisual', {
            sections: [section],
            // to make sure its above a gridhelper if there is any
            position: { x: 0, y: 0.001, z: 0 },
        });
        maxwhere_1.wom.render(this.manualObj);
    }
    clear() {
        this.manualObj.clear();
    }
}
exports.default = AxesHelper;
